### PR TITLE
Us121 note expansion - The ClientData PR

### PR DESCRIPTION
**Pull request by aneuhold**
### Overview

This pull request implements US121, a checkbox, and a bunch of performance improvements through the use of the `ClientData` class / file.

Gotta go fast

`ClientData` is built to circumvent the constant cascading re-renders that React was doing when we held state in the top `App` component. See below for an idea of how this works:

ClientData logic

For an example of how this listener setup / teardown looks, see below for an example from `ProjectTable`:

```js
  const listenerId = `ProjectTable`;
  /**
   * Subscribe to changes in the projects array on the user.
   */
  useEffect(() => {
    ClientData.addUserPropertyListener(listenerId, 'projects', () => {
      const newProjects = [...ClientData.getUser().projects];
      setProjectIds(newProjects);
    });

    return function cleanup() {
      ClientData.removeUserPropertyListener('projects', listenerId);
    };
  }, []);
```

That last little `[]` as the second argument to the `useEffect` function makes it so this function runs once when the component is mounted, and runs the `cleanup` function when the component is unmounted. It doesn't need to be named `cleanup` though, it can be anonymous. See also [this link from the React documentation](https://reactjs.org/docs/hooks-effect.html#tip-optimizing-performance-by-skipping-effects) for more info on how that works. Primarily the yellow note below where that links to. They can explain better than I can. 

### Includes:

- client/src/main/App.tsx 
- client/src/main/components/CompletableRow/CompletableRow.tsx 
- client/src/main/components/CompletableRow/CompletedCheckbox.tsx 
- client/src/main/components/CompletableRow/DateInput.tsx 
- client/src/main/components/CompletableRow/NoteButton.tsx 
- client/src/main/components/NoteInput.tsx → ...n/components/CompletableRow/NoteInput.tsx 
- components/TaskExpanderButton.tsx → ...nts/CompletableRow/TaskExpanderButton.tsx 
- client/src/main/components/DateInput.tsx 
- client/src/main/components/NewItemMenu.tsx 
- client/src/main/components/PriorityButton/PriorityButton.tsx 
- client/src/main/components/PriorityButton/PriorityDialog.tsx 
- client/src/main/components/PriorityButton/PriorityInput.tsx 
- client/src/main/components/ProjectRow.tsx 
- client/src/main/components/ProjectTable.tsx 
- client/src/main/components/SettingsDialog/SettingsDialog.tsx 
- client/src/main/components/SimpleTextInput.tsx 
- client/src/main/components/SortInput.tsx 
- client/src/main/components/TaskMenu/TaskMenu.tsx 
- client/src/main/components/TaskRow.tsx 
- client/src/main/components/TopMenuBar.tsx 
- client/src/main/logic/ClientData.ts 
- client/src/main/logic/comparisonFunctions.ts 
- client/src/main/logic/dbTypes.ts 
- client/src/main/logic/fetchMethods.ts 
- client/src/main/logic/sortingFunctions.ts 
- node_server/src/main/models/project.ts 
- node_server/src/main/models/task.ts 
- node_server/src/main/models/user.ts 
- node_server/src/main/routes/users.ts 
- node_server/src/main/swagger.ts 

### Developer Checklist:

- [x] My code compiles and runs locally
- [x] The code follows the `QualityPolicy.md` file
- [x] My code has been developer-tested and includes unit tests
- [x] I have considered proper use of exceptions
- [x] I have eliminated IDE warnings

### Refs:

- [US 121](https://tree.taiga.io/project/aneuhold-pointspire/us/121?milestone=269364)

### CODE REVIEW

**Comment by NathanJGaul** [comment]
This branch broke github auth somehow. Digging into it right now.

File: client/src/main/App.tsx [hunk 1]
```diff
@@ -277,52 +135,21 @@ class App extends React.Component<AppProps, AppState> {
...
             githubClientId={githubClientId}
-            baseServerUrl={baseServerUrl}
-            sendUpdatedUserToServer={sendUpdatedUserToServer}
             alert={alert}
-            userSettings={user ? user.settings : undefined}
-            updateSettings={updateSettings}
             appTheme={appTheme}
             setTheme={setTheme}
-            loggedIn={!!user}
+            loggedIn={!!ClientData.getUser()}
```
Comment (NathanJGaul) [inline client/src/main/App.tsx:150]: I believe the TopMenuBar is not getting 'notified' of the change in login state from ClientData.

Comment (aneuhold) [reply client/src/main/App.tsx:150]: Good catch. Fixing that now. Thank you

Comment (NathanJGaul) [reply client/src/main/App.tsx:150]: Glad you are looking into it, trying to fix this without fully understanding the ClientData changes was making me go around in circles.

Comment (aneuhold) [reply client/src/main/App.tsx:150]: To fix it, it looks like adding the capability to add listeners to the user object itself will be a clean option. Then the TopMenuBar will be a listener of changes to that and update it's `loggedIn` property accordingly. Going to try that now.

**Comment by NathanJGaul** [comment]
### Review
This PR implements a 'state' outside of react that has proven more customizable with increased performance on the previous iteration with just react's built-in state at the cost of some complexity. After reading through and understanding the changes I see the benefits in performance. 

At the moment, these changes do not build on Netlify due to eslint errors that show on Netlify but not in the local dev environment. A userstoryless task will be added to implement Netlify Dev, a local Netlify test environment, that hopefully help us solve this issue before merging to master.

### Reviewer Checklist
- [x] Comments are comprehensible and add something to the maintainability of the code (trivial comments do not help, rather write readable code). Also JSDoc comments are present. 
- [x] The pull request is a fast forward for the branch being merged into
- [x] Types have been generalized where possible
- [x] Exceptions have been used appropriately
- [x] Repetitive code has been factored out
- [x] Frameworks have been used appropriately – methods have all been defined appropriately
- [x] Unit tests are present and correct
- [ ] Tests have been run locally to make sure they pass
- [x] Code fulfills acceptance criteria for the User Story

**Comment by NathanJGaul** [comment]
node_server test failure:
```
1) PointSpire tests
       /api/users
         POST /id/projects
           should add a project if valid content is sent:
     Error: Timeout of 2000ms exceeded. For async tests and hooks, ensure "done()" is called; if returning a Promise, ensure it resolves. (/home/nathanjgaul/Documents/projects/PointSpire/node_server/src/test/testSuite.ts)

```

**Comment by aneuhold** [comment]
@NathanJGaul that error doesn't seem to be coming up on this end. But I did notice a similar issue this past weekend that went away after running the tests a second time. Maybe the mLab server spins up after not being used for a while?

**Comment by NathanJGaul** [comment]
> node_server test failure:
> 
> ```
> 1) PointSpire tests
>        /api/users
>          POST /id/projects
>            should add a project if valid content is sent:
>      Error: Timeout of 2000ms exceeded. For async tests and hooks, ensure "done()" is called; if returning a Promise, ensure it resolves. (/home/nathanjgaul/Documents/projects/PointSpire/node_server/src/test/testSuite.ts)
> ```

Issue was on my network

**Pull request merged**